File: Ijar-app/Ijar/Property.swift
import Foundation

struct Property: Identifiable {
    let id: String
    let images: [String]
    let price: String
    let bedrooms: Int
    let bathrooms: Int
    let address: String
    let area: String
    let rightmoveUrl: String?
    let agentPhone: String?
    let agentName: String?
    let branchName: String?
    let latitude: Double?
    let longitude: Double?

    // Property details (fetched on-demand)
    var description: String?
    var keyFeatures: [String]?
    var propertyType: String?
    var floorArea: String?
    var epcRating: String?
    var councilTaxBand: String?
    var tenure: String?
    var listingDate: String?
    var availableFrom: String?
    var floorplanImages: [String]?

    /// Returns "Studio" if bedrooms is 0, otherwise returns the number as a string
    var bedroomText: String {
        bedrooms == 0 ? "Studio" : "\(bedrooms)"
    }

    init(id: String = UUID().uuidString, images: [String], price: String, bedrooms: Int, bathrooms: Int, address: String, area: String, rightmoveUrl: String? = nil, agentPhone: String? = nil, agentName: String? = nil, branchName: String? = nil, latitude: Double? = nil, longitude: Double? = nil, description: String? = nil, keyFeatures: [String]? = nil, propertyType: String? = nil, floorArea: String? = nil, epcRating: String? = nil, councilTaxBand: String? = nil, tenure: String? = nil, listingDate: String? = nil, availableFrom: String? = nil, floorplanImages: [String]? = nil) {
        self.id = id
        self.images = images
        self.price = price
        self.bedrooms = bedrooms
        self.bathrooms = bathrooms
        self.address = address
        self.area = area
        self.rightmoveUrl = rightmoveUrl
        self.agentPhone = agentPhone
        self.agentName = agentName
        self.branchName = branchName
        self.latitude = latitude
        self.longitude = longitude
        self.description = description
        self.keyFeatures = keyFeatures
        self.propertyType = propertyType
        self.floorArea = floorArea
        self.epcRating = epcRating
        self.councilTaxBand = councilTaxBand
        self.tenure = tenure
        self.listingDate = listingDate
        self.availableFrom = availableFrom
        self.floorplanImages = floorplanImages
    }
}

File: backend/src/models/property.ts
import { RightmoveProperty } from '../scraper-types';
import { RightmoveAPIProperty } from '../rightmove-api';

export interface LiveSearchProperty {
  id: string;
  images: string[];
  price: string;
  bedrooms: number;
  bathrooms: number;
  address: string;
  area: string;
  rightmoveUrl: string;
  agentPhone: string | null;
  agentName: string | null;
  branchName: string | null;
  latitude: number | null;
  longitude: number | null;
  summary?: string;
  propertyType?: string;
}

// Transform from scraper response (legacy)
export function transformProperty(property: RightmoveProperty): LiveSearchProperty {
  const price = property.price?.displayPrices?.[0]?.displayPrice ||
                property.price?.displayPrice ||
                'Price on request';

  const addressParts = property.displayAddress.split(',');
  const area = addressParts.length > 1
    ? addressParts[addressParts.length - 1].trim()
    : '';

  const images = property.images?.map(img => img.srcUrl || img.url).filter(Boolean) || [];

  return {
    id: property.id.toString(),
    images,
    price,
    bedrooms: property.bedrooms || 0,
    bathrooms: property.bathrooms || 0,
    address: property.displayAddress,
    area,
    rightmoveUrl: `https://www.rightmove.co.uk${property.propertyUrl}`,
    agentPhone: property.customer?.contactTelephone || null,
    agentName: property.customer?.brandTradingName || null,
    branchName: property.customer?.branchName || null,
    latitude: property.location?.latitude || null,
    longitude: property.location?.longitude || null
  };
}

// Transform from API response (new)
export function transformAPIProperty(property: RightmoveAPIProperty): LiveSearchProperty {
  const price = property.displayPrices?.[0]?.displayPrice ||
                `£${property.monthlyRent} pcm` ||
                'Price on request';

  const addressParts = property.address.split(',');
  const area = addressParts.length > 1
    ? addressParts[addressParts.length - 1].trim()
    : '';

  // Get images from thumbnailPhotos, upgrading to larger size
  const images = property.thumbnailPhotos?.map(photo => {
    // Replace max_656x437 with larger size for better quality
    return photo.url.replace('max_656x437', 'max_1024x682');
  }) || [];

  // If no thumbnailPhotos, use the main thumbnail
  if (images.length === 0 && property.photoLargeThumbnailUrl) {
    images.push(property.photoLargeThumbnailUrl.replace('max_656x437', 'max_1024x682'));
  }

  return {
    id: property.identifier.toString(),
    images,
    price,
    bedrooms: property.bedrooms || 0,
    bathrooms: 0, // API listing doesn't include bathrooms, will get from details
    address: property.address.trim(),
    area,
    rightmoveUrl: `https://www.rightmove.co.uk/properties/${property.identifier}`,
    agentPhone: property.branch?.contactTelephoneNumber || null,
    agentName: property.branch?.brandName || null,
    branchName: property.branch?.name || null,
    latitude: property.latitude || null,
    longitude: property.longitude || null,
    summary: property.summary || undefined,
    propertyType: property.propertyType || undefined
  };
}
